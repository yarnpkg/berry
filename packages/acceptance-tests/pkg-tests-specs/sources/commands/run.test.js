describe(`Commands`, () => {
  for (const [description, args] of [[`with prefix`, [`run`]], [`without prefix`, []]]) {
    describe(`run ${description}`, () => {
      test(`it should run the selected script if available`, makeTemporaryEnv({
        scripts: {
          foo: `echo hello`,
        },
      }, async ({path, run, source}) => {
        await run(`install`);

        await expect(run(...args, `foo`)).resolves.toMatchObject({
          stdout: `hello\n`,
        });
      }));

      test(`it should properly forward the script exit codes`, makeTemporaryEnv({
        scripts: {
          foo: `exit 42`,
        },
      }, async ({path, run, source}) => {
        await run(`install`);

        await expect(run(...args, `foo`)).rejects.toMatchObject({
          code: 42,
        });
      }));
    });
  }

  describe(`run misc`, () => {
    test(
      `it should run the binaries from our dependencies if available`,
      makeTemporaryEnv(
        {
          dependencies: {
            [`has-bin-entries`]: `1.0.0`,
          },
        },
        async ({path, run, source}) => {
          await run(`install`);

          await expect(run(`run`, `has-bin-entries`, `success`)).resolves.toMatchObject({
            stdout: `success\n`,
          });
        },
      ),
    );

    test(
      `it should prefer scripts over binaries`,
      makeTemporaryEnv(
        {
          dependencies: {
            [`has-bin-entries`]: `1.0.0`,
          },
          scripts: {
            [`has-bin-entries`]: `echo hello world`,
          },
        },
        async ({path, run, source}) => {
          await run(`install`);

          await expect(run(`run`, `has-bin-entries`)).resolves.toMatchObject({
            stdout: `hello world\n`,
          });
        },
      ),
    );

    test(
      `it should ignore scripts when --binaries-only is set`,
      makeTemporaryEnv(
        {
          dependencies: {
            [`has-bin-entries`]: `1.0.0`,
          },
          scripts: {
            [`has-bin-entries`]: `echo hello world`,
          },
        },
        async ({path, run, source}) => {
          await run(`install`);

          await expect(run(`run`, `--binaries-only`, `has-bin-entries`, `success`)).resolves.toMatchObject({
            stdout: `success\n`,
          });
        },
      ),
    );

    test(
      `it shouldn't require the "--" flag to stop interpreting options after "run" commands (scripts)`,
      makeTemporaryEnv(
        {
          scripts: {
            [`hello`]: `echo`,
          },
        },
        async ({path, run, source}) => {
          await run(`install`);

          await expect(run(`run`, `hello`, `--hello`)).resolves.toMatchObject({
            stdout: `--hello\n`,
          });
        },
      ),
    );

    test(
      `it shouldn't require the "--" flag to stop interpreting options after "run" commands (binaries)`,
      makeTemporaryEnv(
        {
          dependencies: {
            [`has-bin-entries`]: `1.0.0`,
          },
        },
        async ({path, run, source}) => {
          await run(`install`);

          await expect(run(`run`, `has-bin-entries`, `--hello`)).resolves.toMatchObject({
            stdout: `--hello\n`,
          });
        },
      ),
    );
    test(`it should print the list of available scripts if no parameters passed to command`,
      makeTemporaryEnv(
        {
          scripts: {
            foo: `echo hello`,
            bar: `echo hi`,
          },
        },
        async ({path, run, source}) => {
          const {code, stdout, stderr} = await run(`run`);
          expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );
  });
});
