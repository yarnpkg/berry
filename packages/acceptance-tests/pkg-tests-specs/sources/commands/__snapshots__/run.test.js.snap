// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands run misc it should print the list of available scripts if no parameters passed to command 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: foo   'echo hello'
➤ YN0000: bar   'echo hi'
➤ YN0000: Done
",
}
`;
